Apps/FlightSoftware/Watchdog2/src/i2c.c

```diff
@@ -0,0 +1,318 @@
+    I2C_Master_WriteReg(I2C_SLAVE_ADDR, CONTROL, &fuel_gauge_write_control_reg, 1);
+}
+
+//******************************************************************************
+// I2C Interrupt ***************************************************************
+//******************************************************************************
+
+#if defined(__TI_COMPILER_VERSION__) || defined(__IAR_SYSTEMS_ICC__)
+#pragma vector = USCI_B0_VECTOR
+__interrupt void USCI_B0_ISR(void)
+#elif defined(__GNUC__)
+void __attribute__ ((interrupt(USCI_B0_VECTOR))) USCI_B0_ISR (void)
+#else
+#error Compiler not supported!
+#endif
+{
+  //Must read from UCB0RXBUF
+  uint8_t rx_val = 0;
+  switch(__even_in_range(UCB0IV, USCI_I2C_UCBIT9IFG))
+  {
+    case USCI_NONE:          break;         // Vector 0: No interrupts
+    case USCI_I2C_UCALIFG:   break;         // Vector 2: ALIFG
+    case USCI_I2C_UCNACKIFG:                // Vector 4: NACKIFG
+      break;
+    case USCI_I2C_UCSTTIFG:  break;         // Vector 6: STTIFG
+    case USCI_I2C_UCSTPIFG:  break;         // Vector 8: STPIFG
+    case USCI_I2C_UCRXIFG3:  break;         // Vector 10: RXIFG3
+    case USCI_I2C_UCTXIFG3:  break;         // Vector 12: TXIFG3
+    case USCI_I2C_UCRXIFG2:  break;         // Vector 14: RXIFG2
+    case USCI_I2C_UCTXIFG2:  break;         // Vector 16: TXIFG2
+    case USCI_I2C_UCRXIFG1:  break;         // Vector 18: RXIFG1
+    case USCI_I2C_UCTXIFG1:  break;         // Vector 20: TXIFG1
+    case USCI_I2C_UCRXIFG0:                 // Vector 22: RXIFG0
+        rx_val = UCB0RXBUF;
+        if (RXByteCtr)
+        {
+          ReceiveBuffer[ReceiveIndex++] = rx_val;
+          RXByteCtr--;
+        }
+
+        if (RXByteCtr == 1)
+        {
+          UCB0CTLW0 |= UCTXSTP;
+        }
+        else if (RXByteCtr == 0)
+        {
+          UCB0IE &= ~UCRXIE;
+          MasterMode = IDLE_MODE;
+//          __bic_SR_register_on_exit(CPUOFF);      // Exit LPM0
+        }
+        break;
+    case USCI_I2C_UCTXIFG0:                 // Vector 24: TXIFG0
+        switch (MasterMode)
+        {
+          case TX_REG_ADDRESS_MODE:
+              UCB0TXBUF = TransmitRegAddr;
+              if (RXByteCtr)
+                  MasterMode = SWITCH_TO_RX_MODE;   // Need to start receiving now
+              else
+                  MasterMode = TX_DATA_MODE;        // Continue to transmision with the data in Transmit Buffer
+              break;
+
+          case SWITCH_TO_RX_MODE:
+              UCB0IE |= UCRXIE;              // Enable RX interrupt
+              UCB0IE &= ~UCTXIE;             // Disable TX interrupt
+              UCB0CTLW0 &= ~UCTR;            // Switch to receiver
+              MasterMode = RX_DATA_MODE;    // State state is to receive data
+              UCB0CTLW0 |= UCTXSTT;          // Send repeated start
+              if (RXByteCtr == 1)
+              {
+                  //Must send stop since this is the N-1 byte
+                  while((UCB0CTLW0 & UCTXSTT));
```

Comment:
How confident are we that this will happen _very_ quickly? If we're not _very_ confident, then it shouldn't be in an ISR

Reply:
@SchwartzCode

Reply:
Based on page 4 of the [data sheet for the component](https://www.analog.com/media/en/technical-documentation/data-sheets/2944fa.pdf) this file is communicating with, the largest minimum hold time for a repeated start condition is 600 ns, which is just under 100 clock cycles for our 16 MHz MSP430. Is that too long?

Would it make more sense to move this loop to a function, and have the ISR just trigger a flag that tells us to call that function? Although that could throw the i2c timing out of wack?